internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// tlsstate manages TLS keypairs on behalf of the daemon's HTTPS server.
+package tlsstate
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha512"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base32"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/fs"
+	"math/big"
+	"os"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	// systemTime can be faked during testing.
+	systemTime = time.Now
+
+	// certificateValidityPeriod determines how long a newly generated
+	// TLS keypair is valid for.
+	certificateValidityPeriod = (10 * 365 * 24 * time.Hour)
+)
+
+type TLSManager struct {
+	// tlsDir is the location of the PEM keypair files.
+	tlsDir         string
+	tlsKeyPairLock sync.Mutex
+	tlsKeyPairs    []*TLSKeyPair
+}
+
+// NewManager creates a new TLS manager and loads existing keypairs from the
+// tlsDir directory.
+//
+// The TLS manager will generate keypairs on demand, so no external keypair
+// generation or lifecycle management of existing keypairs will be required.
+//
+// However, is possible to pre-seed the tlsDir with TLSv1.3 supported keypairs.
+// The following steps provide an example of how to create a TLSv1.3 supported
+// self-signed TLS keypair (the private key must be encoded in PKCS8 format):
+//
+// ~/> openssl genpkey -algorithm EC -pkeyopt ec_paramgen_curve:prime256v1 -out ec-p256.pem
+// ~/> openssl req -new -x509 -key ec-p256.pem -out ec-p256-cert.pem -days 365
+//
+// The name of the file must match the SHA512/384 hash of the raw public certificate
+// encoded in base32 (without padding), and contain both the private and certificate
+// blocks in a single PEM file:
+//
+// ~/> FINGERPRINT=$(openssl x509 -in ec-p256-cert.pem -outform DER | openssl dgst -sha384 -binary | base32 -w0 | tr -d '=')
+// ~/> cat ec-p256.pem > ${FINGERPRINT}.pem
+// ~/> cat ec-p256-cert.pem >> ${FINGERPRINT}.pem
+//
+// The file permission must be 0o600:
+//
+// ~/> chmod 600 ${FINGERPRINT}.pem
+//
+// The file ${FINGERPRINT}.pem should be copied into tlsDir before the manager is started.
+func NewManager(tlsDir string) (*TLSManager, error) {
+	manager := &TLSManager{tlsDir: tlsDir}
+	if err := manager.loadTLSKeyPairs(); err != nil {
+		return nil, err
+	}
+	return manager, nil
+}
+
+// TLSKeyPair returns the currently preferred keypair. This method may generate
+// a suitable TLS keypair if none exists as the time of the call. The
+// generated keypair will be installed (persisted) in the TLS directory.
+func (m *TLSManager) TLSKeyPair() (*TLSKeyPair, error) {
```

Comment:
```suggestion
func (m *TLSManager) PreferredKeyPair() (*TLSKeyPair, error) {
```

Reply:
PR reworked. N/A for current PR. Resolving.